### PR TITLE
REF: Add ExtensionArray.map

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes https://github.com/pandas-dev/pandas/issues/23179

Working with the `.map` method is very uncomfortable in main, because of issues with `ExtensionArray`:
* `ExtensionArray` is not required to  have a `.map` method, but may have it.
* Some `ExtensionArray` subclasses, that do have a `map` method have one function parameter (`mapper`), while other have two parameters (`mapper` and `na_action`).
* There is fallback functionality for calculation  `ExtensionArray.map` in `IndexOps.Mixin._map_values`. This means we can calculate `.map` for `ExtensionArrays` indirectly, if we wrap them in a `Series` or `Index`, but we cannot rely on the ability to do  it directly on the `ExtensionArray`.
* `ExtensionArray` subclasses often want create their own `.map` methods
* Because of all the above issues it is often very maze-like to follow code that uses `.map`

This PR solves this by:
1. moving the functionality in `IndexOpsMixin._map_values` to a function `map_array` in `algorithms`.
2.  adding a `.map` method to `ExtensionArray` that calls the new `map_array` function. Subclasses will override this method where needed.
3. Call the new `map_array` function inside `IndexOpsMixin._map_values`. 
4. Ensuring a common signature for the `.map` method for `ExtensionArray`, i.e. always two method parameters (`mapper` and `na_action`), so we can reliably call the `.map` method on any `ExtensionArray` subclass.

Note that this PR doesn't really add new functionality, because we're mostly just moving code around, though it is now possible to call the `map` method on any `Extensionarray` directly, instead of going through e.g. `Series.map`.

In the main branch, there are currently several `ExtensionArray` subclasses where `na_action='ignore'` doesn't work. I'm working on making it work with all of the `ExtensionArray` subclasses and there is already #51645, which handles `Categorical`/`CategoricalIndex` and more will follow.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Does this close any of #23179, #15706, #32815?

**Comment by topper-123** [comment]
> Does this close any of #23179, #15706, #32815?

* #23179: Yes (adds `.map` to `ExtensionArray`)
* #15706: I don't think this issue is solve while maintaining the performance benefits of a dedicated `Categorical.map` , (i.e. doing the op only once per category is intended).
* #32815: not in this PR, but that is fixed by #51645 for the `Categorical`/`CategoricalIndex` case. I'll take on `DateTimeArray`/`DateTimeIndex` etc. in a follow-up, so that will be fixed shortly.

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -1678,3 +1679,73 @@ def union_with_duplicates(
...
+
+        # Since values were input this means we came from either
+        # a dict or a series and mapper should be an index
+        indexer = mapper.index.get_indexer(arr)
+        new_values = take_nd(mapper._values, indexer)
+
+        return new_values
+
+    # we must convert to python types
+    values = arr.astype(object)
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:1747]: copy=False?

**Comment by jbrockmendel** [comment]
Might need something in the extension tests?

File: pandas/core/algorithms.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/algorithms.py:1747]: Yeah, good point.

**Comment by topper-123** [comment]
> Might need something in the extension tests?

I've added some simple tests in the newest version.

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -1678,3 +1679,73 @@ def union_with_duplicates(
         unique_vals = ensure_wrapped_if_datetimelike(unique_vals)
     repeats = final_count.reindex(unique_vals).values
     return np.repeat(unique_vals, repeats)
+
+
+def map_array(arr: ArrayLike, mapper, na_action=None):
```
Comment (mroeschke) [inline pandas/core/algorithms.py:1684]: Could you type na_action here?

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -1678,3 +1679,73 @@ def union_with_duplicates(
...
+        a MultiIndex will be returned.
+    """
+    if na_action not in (None, "ignore"):
+        msg = f"na_action must either be 'ignore' or None, {na_action} was passed"
+        raise ValueError(msg)
+
+    # we can fastpath dict/Series to an efficient map
+    # as we know that we are not going to have to yield
+    # python types
+    if is_dict_like(mapper):
```
Comment (mroeschke) [inline pandas/core/algorithms.py:1710]: Confirming that `na_action="ignore"` is not relevant for this branch?

Comment (topper-123) [reply pandas/core/algorithms.py:1710]: Yes, this branch converts the mapper to either a Series or a callable, so `na_action` isn't relevant here, but will be used later in the function.

File: pandas/core/algorithms.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/algorithms.py]: I've pushed a new version with typed parameters.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -77,27 +77,6 @@ repos:
             --linelength=88,
             '--filter=-readability/casting,-runtime/int,-build/include_subdir,-readability/fn_size'
         ]
--   repo: https://github.com/pycqa/pylint
```
Comment (mroeschke) [inline .pre-commit-config.yaml:80]: Don't think this should be removed?

Comment (topper-123) [reply .pre-commit-config.yaml]: yes... I remove pylint temporary when running `pre-commit run --color=always --hook-stage manual --all-files`, but I should find a better way.

I've updated a new version again.

**Comment by topper-123** [comment]
I've updated.

**Review by mroeschke — APPROVED** [review]
The implementation LGTM

**Comment by topper-123** [comment]
Gentle ping.

I’m intending to follow up with fixes towards #51644, and it would be great if this could be merged.

**Comment by mroeschke** [comment]
Merge when ready @jbrockmendel

**Review by jbrockmendel — APPROVED** [review]

**Comment by jbrockmendel** [comment]
thanks @topper-123

**Comment by jbrockmendel** [comment]
@topper-123 does any of your recent work on this close #15706?

**Comment by topper-123** [comment]
I'vew answered in #15706 to keep the discussion in one place.

**Pull request merged**